src/server/pkg/tarutil/util.go

```diff
@@ -0,0 +1,251 @@
+package tarutil
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/tar"
+)
+
+type File interface {
+	Header() (*tar.Header, error)
+	Content(io.Writer) error
+}
+
+type file struct {
+	hdr  *tar.Header
+	data []byte
+}
+
+func NewFile(name string, data []byte) File {
+	return newFile(NewHeader(name, int64(len(data))), data)
+}
+
+func newFile(hdr *tar.Header, data []byte) File {
+	return &file{
+		hdr:  hdr,
+		data: data,
+	}
+}
+
+func NewHeader(name string, size int64) *tar.Header {
+	return &tar.Header{
+		Name: name,
+		Size: size,
+	}
+}
+
+func (f *file) Header() (*tar.Header, error) {
+	return f.hdr, nil
+}
+
+func (f *file) Content(w io.Writer) error {
+	_, err := w.Write(f.data)
+	return err
+}
+
+type streamFile struct {
+	hdr *tar.Header
+	r   io.Reader
+}
+
+func NewStreamFile(name string, size int64, r io.Reader) File {
+	return newStreamFile(NewHeader(name, size), r)
+}
+
+func newStreamFile(hdr *tar.Header, r io.Reader) File {
+	return &streamFile{
+		hdr: hdr,
+		r:   r,
+	}
+}
+
+func (sf *streamFile) Header() (*tar.Header, error) {
+	return sf.hdr, nil
+}
+
+func (sf *streamFile) Content(w io.Writer) error {
+	_, err := io.Copy(w, sf.r)
+	return err
+}
+
+func WriteFile(tw *tar.Writer, file File) error {
+	hdr, err := file.Header()
+	if err != nil {
+		return err
+	}
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+	if err := file.Content(tw); err != nil {
+		return err
+	}
+	return tw.Flush()
+}
+
+func WithWriter(w io.Writer, cb func(*tar.Writer) error) (retErr error) {
+	tw := tar.NewWriter(w)
+	defer func() {
+		if retErr == nil {
+			retErr = tw.Close()
+		}
+	}()
+	return cb(tw)
+}
+
+func Iterate(r io.Reader, cb func(File) error, stream ...bool) error {
+	tr := tar.NewReader(r)
+	for {
+		hdr, err := tr.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return err
+		}
+		if len(stream) > 0 && stream[0] {
+			if err := cb(newStreamFile(hdr, tr)); err != nil {
+				return err
+			}
+			continue
+		}
+		buf := &bytes.Buffer{}
+		_, err = io.Copy(buf, tr)
+		if err != nil {
+			return err
+		}
+		if err := cb(newFile(hdr, buf.Bytes())); err != nil {
+			return err
+		}
+	}
+}
+
+func Equal(file1, file2 File, full ...bool) (bool, error) {
+	buf1, buf2 := &bytes.Buffer{}, &bytes.Buffer{}
```

Comment:
use a `hash.Hash` for this.  It implements `io.Writer`.

Reply:
I would prefer to leave it as is for now. We can revisit this later if we want to use this for checking if large files are equal.